src/data/list/duplicate.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2018 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky, Chris Hughes
+-/
+import data.list.basic
+
+/-!
+# List duplicates
+
+## Main definitions
+
+* `list.duplicate x l : Prop` is an inductive property that holds when `x` is a duplicate in `l`
+
+-/
+
+variable {α : Type*}
+
+namespace list
+
+/-- Property that an element `x : α` of `l : list α` can be found in the list more than once. -/
+inductive duplicate (x : α) : list α → Prop
+| cons_mem {l : list α} : x ∈ l → duplicate (x :: l)
+| cons_duplicate {y : α} {l : list α} : duplicate l → duplicate (y :: l)
+
+local infix ` ∈+ `:1 := list.duplicate
+
+variables {l : list α} {x : α}
+
+lemma mem.duplicate_cons_self (h : x ∈ l) : x ∈+ x :: l := duplicate.cons_mem h
+
+lemma duplicate.duplicate_cons (h : x ∈+ l) (y : α) : x ∈+ y :: l := duplicate.cons_duplicate h
+
+lemma duplicate.mem (h : x ∈+ l) : x ∈ l :=
+begin
+  induction h with l' h y l' h hm,
+  { exact mem_cons_self _ _ },
+  { exact mem_cons_of_mem _ hm }
+end
+
+lemma duplicate.mem_cons_self (h : x ∈+ x :: l) : x ∈ l :=
+begin
+  cases h with _ h _ _ h,
+  { exact h },
+  { exact h.mem }
+end
+
+@[simp] lemma duplicate_cons_self_iff : (x ∈+ x :: l) ↔ x ∈ l :=
+⟨duplicate.mem_cons_self, mem.duplicate_cons_self⟩
+
+lemma duplicate.ne_nil (h : x ∈+ l) : l ≠ [] :=
+λ H, (mem_nil_iff x).mp (H ▸ h.mem)
+
+@[simp] lemma not_duplicate_nil (x : α) : ¬ (x ∈+ []) :=
+λ H, by simpa using H.ne_nil
+
+lemma duplicate.ne_singleton (h : x ∈+ l) (y : α) : l ≠ [y] :=
+begin
+  induction h with l' h z l' h hm,
+  { simp [ne_nil_of_mem h] },
+  { simp [ne_nil_of_mem h.mem] }
+end
+
+@[simp] lemma not_duplicate_singleton (x y : α) : ¬ (x ∈+ [y]) :=
+λ H, by simpa using H.ne_singleton _
+
+lemma duplicate.eq_and_mem_or_duplicate {y : α} (h : x ∈+ y :: l) : (y = x ∧ x ∈ l) ∨ (x ∈+ l) :=
```

Comment:
turn into an `iff` or add corresponding `iff` (that should also signifincantly shorten `decidable_duplicate`). And `duplicate_cons_iff_of_ne` is just a special case of this.